### ISSUE TITLE
Document Schema metadata expectations

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#12736]
### Is your feature request related to a problem or challenge?

There is an (implicit) assumption that metadata attached to Schema is preserved during certain operations in DataFusion.

However, this expectation is clearly not well tested or documented (e.g. see https://github.com/apache/datafusion/issues/12733)

### Describe the solution you'd like

I would like the assumptions documented

### Describe alternatives you've considered

I suggest documentation on in https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.LogicalPlan.html that explains the high level assumptions

Then add a note /link to that section from the optimizers:
https://docs.rs/datafusion/latest/datafusion/optimizer/trait.AnalyzerRule.html
https://docs.rs/datafusion/latest/datafusion/optimizer/trait.OptimizerRule.html
https://docs.rs/datafusion/latest/datafusion/physical_optimizer/trait.PhysicalOptimizerRule.html

My understanding of the high level assumptions are:
* schema level metadata: always passed through
* field level metadata: when there is a clear 1-1 correspondence from an input column with metadata to an output column, the metadata should be preserved

Examples
* `PROJECT(a, b+c)` --> field metadata on`a` should be preserved, no field metadata on `b+c`
* `SUM(a) .. GROUP BY b` --> field metadata on `b` is preserved, not on `a`

**Comment by wiedld** [comment]
take

**Comment by alamb** [comment]
I will take a shot at documenting this

**Comment by alamb** [comment]
https://github.com/apache/datafusion/pull/13305#issuecomment-2466242845 has some additional context

**Comment by westonpace** [comment]
I'm not sure if this is related or not but I encountered an error during optimization:

```
Error: join_selection caused by Internal error: PhysicalOptimizer rule 'join_selection' failed, due to generate a different schema, original schema: Schema { fields: [Field { name: "a", ...  }], metadata: {metadata_from_table_one}, new schema: Schema { fields: [Field { name: "a", ... }], metadata: {metadata_from_table_two}. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

I think the join operator is either returning the metadata from the first or second input and so an optimization that swaps the join order results in the output schema of the join operator changing and this causes the optimizer to bail.

**Comment by alamb** [comment]
That definitely sounds like a bug

**Comment by wiedld** [comment]
@alamb -- I'm taking a look back at this ticket. There has been some changes, and I'm unclear as to long term plans.

There is a new [FieldMetadata](https://github.com/apache/datafusion/blob/420a862faa98abdb941f22d6b7c9bd463a51c777/datafusion/expr/src/expr.rs#L448) which only applies to literals and unifies it's handling? What about non-literal fields?

How (or is) this related to a consolidation of how schema field metadata is handled? Are we unifying a single wrapper abstraction to handle how merging (e.g. union), precedence (e.g. joins?), aggregates, lists, and/or casting is handled?

**Comment by alamb** [comment]
> [@alamb](https://github.com/alamb) -- I'm taking a look back at this ticket. There has been some changes, and I'm unclear as to long term plans.
> 
> There is a new [FieldMetadata](https://github.com/apache/datafusion/blob/420a862faa98abdb941f22d6b7c9bd463a51c777/datafusion/expr/src/expr.rs#L448) which only applies to literals and unifies it's handling? What about non-literal fields?
> 
> How (or is) this related to a consolidation of how schema field metadata is handled? Are we unifying a single wrapper abstraction to handle how merging (e.g. union), precedence (e.g. joins?), aggregates, lists, and/or casting is handled?

TLDR is I don't know of any unified plan. 

What I would recommend doing is starting by documenting what (little) is known today (about projections and aliases)

Then as we evolve and the plans become clearer, we will at least have a place to improve the documenetion

I think starting with something even if it is incomplete, is better than the status quo (nothing)

**Comment by wiedld** [comment]
Non-exhaustive list, as I find things:
- [x] handling of UNION logical plan schemas, fields, and metadata: https://github.com/apache/datafusion/pull/17248
- [x] different definitions of schema & field `merging`, in logical plans & their expressions: https://github.com/apache/datafusion/pull/17249
- [x] LogicalPlan handling of the projection of schema, fields, and metadata: https://github.com/apache/datafusion/pull/17250
- [ ] initial source of schema, fields, and metadata based upon ListingTable and other schema adapters: TODO
- [ ] TODO: assess scope of physical plan handling.

**Comment by adriangb** [comment]
I found a bug with first_value / last_value: https://github.com/apache/datafusion/pull/19335
